Allow collection expressions for 'System.Collections.ObjectModel' collection types

AlexRadch opened the pull request:
Close #110161

AlexRadch committed:
#110161 Allow collection expressions for 'System.Collections.ObjectModel' collection types
https://github.com/dotnet/runtime/issues/110161

huoyaoyuan commented on src/libraries/System.Private.CoreLib/src/System/Collections/ObjectModel/ReadOnlyCollection.cs:
```diff
@@ -229,4 +231,35 @@ void IList.RemoveAt(int index)
+        [EditorBrowsable(EditorBrowsableState.Never)]
+#pragma warning disable IDE0301 // Simplify collection initialization
+        public static ReadOnlyCollection<T> CreateCollection<T>(params ReadOnlySpan<T> values) =>
+            values.Length <= 0 ? ReadOnlyCollection<T>.Empty : new ReadOnlyCollection<T>(values.ToArray());
```
Use `IsEmpty` to check length of span, like `ImmutableArray.Create`.

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Collections/ObjectModel/ReadOnlyCollection.cs:
```diff
@@ -229,4 +231,35 @@ void IList.RemoveAt(int index)
+        /// <param name="values">The span of values to include in the collection.</param>
+        /// <returns>A new <see cref="ReadOnlyCollection{T}"/> containing the specified values.</returns>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+#pragma warning disable IDE0301 // Simplify collection initialization
```
This sounds not be marked with EditorBrowsable

stephentoub replied:
Please remove the pragma

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Collections/ObjectModel/ReadOnlyCollection.cs:
```diff
@@ -229,4 +231,35 @@ void IList.RemoveAt(int index)
+        /// <param name="values">The span of values to include in the collection.</param>
+        /// <returns>A new <see cref="ReadOnlySet{T}"/> containing the specified values.</returns>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+#pragma warning disable IDE0301 // Simplify collection initialization
```
Same comments

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Collections/ObjectModel/ReadOnlyCollection.cs:
```diff
@@ -229,4 +231,35 @@ void IList.RemoveAt(int index)
         }
     }
+
+    public static class ReadOnlyCollection
```
The type needs xml docs

AlexRadch committed:
Code review

Enhance ReadOnlyCollection with documentation and updates

The `EditorBrowsable` attribute is replaced with a `SuppressMessage` attribute to clarify the intended usage of these methods.

Additionally, the condition for checking if input values are empty is updated from `values.Length <= 0` to `values.IsEmpty` for improved readability and performance.

AlexRadch committed:
Remove unused namespace from ReadOnlyCollection.cs

Eliminated the `using System.ComponentModel;` directive, reducing dependencies and potentially simplifying the code structure.

eiriktsarpalis commented on src/libraries/System.Private.CoreLib/src/System/Collections/ObjectModel/ReadOnlyCollection.cs:
```diff
@@ -229,4 +230,34 @@ void IList.RemoveAt(int index)
+        /// <typeparam name="T">The type of elements in the collection.</typeparam>
+        /// <param name="values">The span of values to include in the collection.</param>
+        /// <returns>A new <see cref="ReadOnlyCollection{T}"/> containing the specified values.</returns>
+        [Diagnostics.CodeAnalysis.SuppressMessage("Style", "IDE0301:Simplify collection initialization", Justification = "This method simplifies ReadOnlyCollection<T> initialization, so you cannot simplify collection initialization inside it.")]
```
We typically suppress warnings using `#pragma` declarations. What necessitates putting a suppress attribute on the implementation?

stephentoub replied:
This doesn't need to be suppressed at all. It's not a warning, only a suggestion.

AlexRadch replied:
> This doesn't need to be suppressed at all. It's not a warning, only a suggestion.

I suppressed this suggestion to prevent it from being accidentally applied to the code because it would cause recursion when it was compiled. I can remove the suppression.

eiriktsarpalis commented on src/libraries/System.Private.CoreLib/src/System/Collections/ObjectModel/ReadOnlyCollection.cs:
```diff
@@ -229,4 +230,34 @@ void IList.RemoveAt(int index)
+        /// <returns>A new <see cref="ReadOnlySet{T}"/> containing the specified values.</returns>
+        [Diagnostics.CodeAnalysis.SuppressMessage("Style", "IDE0301:Simplify collection initialization", Justification = "This method simplifies ReadOnlySet<T> initialization, so you cannot simplify collection initialization inside it.")]
+        public static ReadOnlySet<T> CreateSet<T>(params ReadOnlySpan<T> values) =>
+            values.IsEmpty ? ReadOnlySet<T>.Empty : new ReadOnlySet<T>((HashSet<T>)[.. values]);
```
Using the spread operator on sets doesn't appear to produce optimal code:

https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQAwAIpwCwG5nJYBs6AEgIYDOAFgMoCmALgDwAqAfOgGIBOA9gFtaAB3IA7ET3rlgbdgAoAStOAB5MQBsAniPFz0ASzHCAro0oBKZAG9k6e5gDs6ANroAdO8PGzlALr4SAC+BChwJBQ0DCwc3PxComIAsuIm5BpySirq2rpi+kam5lZItkgOZFR0TPqUTOgAvOhi9ADu8oW+7gAy9GIA5ozUFoEVAGZ8UuQAxtTo8qzo9N7eRZZ2DmUVFXWM7gCCwMDy9CMb9iHlDudO6LuBQUA

Consider handrolling the factory?

AlexRadch replied:
> Using the spread operator on sets doesn't appear to produce optimal code:
> 

Using Count is not entirely appropriate because the enumeration might contain duplicates, and in some cases, there could be many of them. This would result in unnecessary memory allocation.

Additionally, I think that the compiler is capable of optimizing while (enumerator.MoveNext()) loops effectively.

AlexRadch replied:
I will replace the loop with a manual, but should I use the Count property?

stephentoub replied:
> Additionally, I think that the compiler is capable of optimizing while (enumerator.MoveNext()) loops effectively.

The C# compiler shouldn't be emitting use of an enumerator here, just as it doesn't when foreach'ing over a span, but for some reason is. @cston, @RikkiGibson?

RikkiGibson replied:
Looks like a roslyn bug. I see some bits in foreach lowering which are handling the span case and "rewriting as `for`", which are missed in the spread element lowering.

https://github.com/dotnet/roslyn/blob/8e6206417f57fff0a67b3e553f25d1e73899d2bd/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ForEachStatement.cs#L60

https://github.com/dotnet/roslyn/blob/8e6206417f57fff0a67b3e553f25d1e73899d2bd/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs#L1402

eiriktsarpalis replied:
> I will replace the loop with a manual, but should I use the Count property?

Presumably you mean `.Length`? The input is a ReadOnlySpan.

eiriktsarpalis replied:
> Yes! I mean the `Length` property. Is it correct to use `Length` here or not? Sometimes a source may contain many duplicates. This will result in unnecessary memory allocation.

Good point. The compiler [doesn't pre-size the HashSet instances either](https://sharplab.io/#v2:CYLg1APgAgDABFAjAFgNwFgBQsGIHQAyAlgHYCOGmWAEgIYDOAFgMoCmALgDynsB8cAN1oAbAK6t6cALxwA2njxwAoiVEBbVgCdaAI2Gs8AJVYAHVrXYAKRABo4iGI4CUAXVRA==) so we should do the same here.

> Should I rewrite the loop to a manual for, or will the compiler error be fixed?

A handwritten factory is fine I think, it doesn't fundamentally change the amount of IL being generated.

AlexRadch committed:
Remove suppression attributes from collection methods

The `CreateCollection<T>` method in the `ReadOnlyCollection` class had its suppression attribute removed, addressing a style warning related to collection initialization while maintaining its functionality.

Similarly, the `CreateSet<T>` method in the `ReadOnlySet` class also had its suppression attribute removed, preserving its original functionality.

mphelt commented on src/libraries/System.Private.CoreLib/src/System/Collections/ObjectModel/ReadOnlyCollection.cs:
```diff
@@ -229,4 +230,32 @@ void IList.RemoveAt(int index)
+        /// <typeparam name="T">The type of elements in the collection.</typeparam>
+        /// <param name="values">The span of values to include in the collection.</param>
+        /// <returns>A new <see cref="ReadOnlySet{T}"/> containing the specified values.</returns>
+        public static ReadOnlySet<T> CreateSet<T>(params ReadOnlySpan<T> values) =>
```
What is the reason for this method to be `ReadOnlyCollection.CreateSet<T>()` instead of `ReadOnlySet.Create<T>()` in a new static class?
The later is much more user friendly.

huoyaoyuan replied:
The builder methods aren't for users. They are for compilers to build collection expressions.
The compiler only requires static method in non-generic type.

AlexRadch replied:
> What is the reason for this method to be `ReadOnlyCollection.CreateSet<T>()` instead of `ReadOnlySet.Create<T>()` in a new static class? The later is much more user friendly.

I could replace `ReadOnlyCollection.CreateSet<T>()` with the `ReadOnlySet.Create<T>()` method if that's more correct.

huoyaoyuan replied:
> I could replace `ReadOnlyCollection.CreateSet<T>()` with the `ReadOnlySet.Create<T>()` method if that's more correct.

That's not what's approved in API review. Questions around the method name should go back to the [original issue](https://github.com/dotnet/runtime/issues/110161).

AlexRadch replied:
> > I could replace `ReadOnlyCollection.CreateSet<T>()` with the `ReadOnlySet.Create<T>()` method if that's more correct.
> 
> That's not what's approved in API review. Questions around the names should go back to the [original issue](https://github.com/dotnet/runtime/issues/110161).

It says that methods can be created as one class or in separate ones... https://github.com/dotnet/runtime/issues/110161#issuecomment-2515391504

> We could add another type (e.g. System.Collections.ObjectModel.Collection **or one non-generic type per collection** (like we do in immutable).

Sergio0694 commented on src/libraries/System.Runtime/ref/System.Runtime.cs:
```diff
@@ -8366,6 +8366,12 @@ void System.Collections.ICollection.CopyTo(System.Array array, int index) { }
     }
+    public static partial class ReadOnlyCollection
+    {
+        public static System.Collections.ObjectModel.ReadOnlyCollection<T> CreateCollection<T>(params System.ReadOnlySpan<T> values) { throw null; }
```
We should have `[EditorBrowsable(Never)]` on both these `Create` methods (same as the approved API surface in https://github.com/dotnet/runtime/issues/108457#issuecomment-2400438835).

AlexRadch replied:
> We should have `[EditorBrowsable(Never)]` on both these `Create` methods (same as the approved API surface in [#108457 (comment)](https://github.com/dotnet/runtime/issues/108457#issuecomment-2400438835)).

It was here but @stephentoub wrote ["This should not be marked with EditorBrowsable"](https://github.com/dotnet/runtime/pull/111093/commits/74f8733b81974cd242898279346e038936fddf21#r1903262956)

[API review](https://github.com/dotnet/runtime/issues/110161#issuecomment-2515391504) do not have this attribute also.

Should I return `[EditorBrowsable(Never)]` attribute to both methods?

Sergio0694 replied:
I thought API review had just left those out accidentally. I don't feel strongly either way, but we should make sure to be consistent with the other factory APIs and any other ones we're adding I assume (eg. the ones in the other issue).

stephentoub replied:
No, it wasn't accidental. These are perfectly valid for a dev to call directly. Please remove them.

mphelt replied:
If those methods are meant to be called directly then based on this [#110161 (comment)](https://github.com/dotnet/runtime/issues/110161#issuecomment-2515391504) it seams more natural to create 1 non-generic type per collection
like it is done for Immutable and Frozen collections.

AlexRadch replied:
> If those methods are meant to be called directly then based on this [#110161 (comment)](https://github.com/dotnet/runtime/issues/110161#issuecomment-2515391504) it seams more natural to create 1 non-generic type per collection like it is done for Immutable and Frozen collections.

I'll create two different classes for these methods like for the Immutable and Frozen collections, and remove the [EditorBrowsable(Never)] attribute.

Sergio0694 replied:
Pretty sure this would have to go back to API review to split it into different classes, as that's not the API shape that was approved.

As for the attribute, if we're removing it, should we also remove it from the approved API surface of the other `Create` methods for the other collection types? Or is there a specific reason only those are hidden?

stephentoub replied:
> As for the attribute, if we're removing it, should we also remove it from the approved API surface of the other Create methods for the other collection types?

Which are you referring to?

AlexRadch replied:
I separated the methods into different classes, renamed them as `Create`, and removed the [EditorBrowsable(Never)] attribute.

Sergio0694 replied:
> Which are you referring to?

I mean the ones on the factory methods approved in https://github.com/dotnet/runtime/issues/108457#issuecomment-2400438835.

stephentoub replied:
> > Which are you referring to?
> 
> I mean the ones on the factory methods approved in [#108457 (comment)](https://github.com/dotnet/runtime/issues/108457#issuecomment-2400438835).

Those are marked "needs work", with the comment "Rather than putting these on CollectionExtensions, we should just add public ctors that accept ReadOnlySpan"

AlexRadch replied:
> Pretty sure this would have to go back to API review to split it into different classes, as that's not the API shape that was approved.

@Sergio0694 API review says
 
* We could add another type (e.g. System.Collections.ObjectModel.Collection **or one non-generic type per collection (like we do in immutable)**.

Were two classes also approved?

eiriktsarpalis replied:
> Were two classes also approved?

No, these are notes about what got discussed during review. The API as approved puts both factories under the `ReadOnlyCollection` class. Could you update your changes to reflect that? Thanks.

AlexRadch replied:
I'll move it back to one class.

AlexRadch committed:
Refactor `(HashSet<T>)[.. values]` to manual loop

AlexRadch committed:
Marked CreateCollection and CreateSet methods in ReadOnlyCollection with the `[EditorBrowsable(EditorBrowsableState.Never)]` attribute to hide them from IntelliSense.

eiriktsarpalis commented on src/libraries/System.Private.CoreLib/src/System/Collections/ObjectModel/ReadOnlyCollection.cs:
```diff
@@ -229,4 +231,43 @@ void IList.RemoveAt(int index)
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public static ReadOnlySet<T> CreateSet<T>(params ReadOnlySpan<T> values)
+        {
+            if (values.IsEmpty)
+                return ReadOnlySet<T>.Empty;
```
Nit: we generally favor using blocks in all branches
```suggestion
            if (values.IsEmpty)
            {
                return ReadOnlySet<T>.Empty;
            }
```

eiriktsarpalis commented on src/libraries/System.Private.CoreLib/src/System/Collections/ObjectModel/ReadOnlyCollection.cs:
```diff
@@ -229,4 +231,43 @@ void IList.RemoveAt(int index)
+                return ReadOnlySet<T>.Empty;
+
+            HashSet<T> hashSet = [];
+            foreach (T value in values)
+                hashSet.Add(value);
```
```suggestion
            foreach (T value in values)
            {
                hashSet.Add(value);
            }
```

eiriktsarpalis committed:
Apply suggestions from code review

eiriktsarpalis reviewed (approved):
Thanks!

eiriktsarpalis committed:
Merge branch 'main' into CollectionExpressionsForObjectModelCollections

AlexRadch committed:
Refactor ReadOnlyCollection and ReadOnlySet APIs

Updated method names in ReadOnlyCollection and ReadOnlySet from CreateCollection and CreateSet to Create for clarity and consistency.

Removed the CreateSet method from ReadOnlyCollection to streamline functionality.
Adjusted documentation comments for improved clarity.

Updated CollectionBuilder attributes to reflect the new method names, ensuring proper linkage for collection creation.

AlexRadch committed:
Updated the `CollectionBuilder` attribute to reference `ReadOnlySet` instead of `ReadOnlyCollection`, reflecting a change in the collection type used in the `ReadOnlySet` class.

AlexRadch committed:
Code Style

AlexRadch committed:
Rename methods in ReadOnlyCollection and ReadOnlySet

Updated `ReadOnlyCollection` to rename `Create` to `CreateCollection` and introduced `CreateSet` for creating `ReadOnlySet` instances from spans. Adjusted `CollectionBuilder` attributes accordingly and removed the static `ReadOnlySet` class, integrating its functionality into the main class. Updated the `ReadOnlyCollection` partial class in `System.Runtime.cs` to maintain API consistency.

AlexRadch committed:
Merge branch 'main' into CollectionExpressionsForObjectModelCollections
